#include "MainWindow.h"
#include "ui_MainWindow.h"

#include <QComboBox>
#include <QHBoxLayout>
#include <QLabel>
#include <QMessageBox>
#include <QTableWidget>
#include <QTableWidgetItem>
#include <QToolButton>
#include <QVBoxLayout>
#include <optional>

#include "AchievementGallery.h"
#include "CustomizationPanel.h"
#include "DashboardWidget.h"
#include "GrowthDashboard.h"
#include "LogBrowser.h"
#include "ShopInterface.h"
#include "../core/ShopItem.h"
#include "../core/LogEntry.h"
#include "TaskView.h"
#include "TutorialManager.h"

/**
 * @brief æ„é€ å‡½æ•°ï¼šåˆ›å»ºæ‰€æœ‰å­ç»„ä»¶å¹¶è¿æ¥ä¸šåŠ¡é€»è¾‘ã€‚
 */
MainWindow::MainWindow(rove::data::UserManager& userManager,
                       rove::data::TaskManager& taskManager,
                       rove::data::AchievementManager& achievementManager,
                       rove::data::LogManager& logManager,
                       rove::data::ShopManager& shopManager,
                       rove::data::InventoryManager& inventoryManager,
                       rove::simulation::SerendipityEngine& serendipityEngine,
                       rove::data::GrowthVisualizer& growthVisualizer,
                       QWidget* parent)
    : QMainWindow(parent)
    , ui(std::make_unique<Ui::MainWindow>())
    , m_userManager(userManager)
    , m_taskManager(taskManager)
    , m_achievementManager(achievementManager)
    , m_logManager(logManager)
    , m_shopManager(shopManager)
    , m_inventoryManager(inventoryManager)
    , m_serendipityEngine(serendipityEngine)
    , m_growthVisualizer(growthVisualizer) {
    ui->setupUi(this);

    // åˆå§‹åŒ–å­ç»„ä»¶
    m_dashboard = new DashboardWidget(this);
    m_taskView = new TaskView(m_taskManager, this);
    m_achievementGallery = new AchievementGallery(m_achievementManager, this);
    m_growthDashboard = new GrowthDashboard(m_growthVisualizer, this);
    m_shopInterface = new ShopInterface(m_shopManager, m_inventoryManager, this);
    m_logBrowser = new LogBrowser(m_logManager, this);
    m_customizationPanel = new CustomizationPanel(m_taskManager, m_achievementManager, m_serendipityEngine, this);
    m_tutorialManager = new TutorialManager(this);

    // å®‰è£…å¸ƒå±€åˆ°å„é¡µé¢
    auto setPage = [](QWidget* page, QWidget* child) {
        auto* layout = new QVBoxLayout(page);
        layout->setContentsMargins(0, 0, 0, 0);
        layout->addWidget(child);
    };
    setPage(ui->pageDashboard, m_dashboard);
    setPage(ui->pageTask, m_taskView);
    setPage(ui->pageAchievement, m_achievementGallery);
    setPage(ui->pageGrowth, m_growthDashboard);
    setPage(ui->pageShop, m_shopInterface);
    setPage(ui->pageLog, m_logBrowser);
    setPage(ui->pageCustom, m_customizationPanel);

    setupNavigation();
    connectSignals();
    setupTrayIcon();
    refreshDashboard();
    showRealtimeNotification(m_tutorialManager->currentHint());
}

MainWindow::~MainWindow() = default;

void MainWindow::setupNavigation() {
    // å¯¼èˆªæŒ‰é’®ä¸å †å é¡µå¯¹åº”
    connect(ui->dashboardBtn, &QPushButton::clicked, [this] { onSectionChanged(0); });
    connect(ui->taskBtn, &QPushButton::clicked, [this] { onSectionChanged(1); });
    connect(ui->achievementBtn, &QPushButton::clicked, [this] { onSectionChanged(2); });
    connect(ui->growthBtn, &QPushButton::clicked, [this] { onSectionChanged(3); });
    connect(ui->shopBtn, &QPushButton::clicked, [this] { onSectionChanged(4); });
    connect(ui->logBtn, &QPushButton::clicked, [this] { onSectionChanged(5); });
    connect(ui->customBtn, &QPushButton::clicked, [this] { onSectionChanged(6); });
    connect(ui->skipTutorialBtn, &QPushButton::clicked, m_tutorialManager, &TutorialManager::skip);

    // çš®è‚¤ä¸å¿ƒæƒ…/å®½æ•åˆ¸/èƒŒåŒ…é¢æ¿
    auto* navLayout = qobject_cast<QVBoxLayout*>(ui->navPanel->layout());
    auto* skinCombo = new QComboBox(ui->navPanel);
    skinCombo->addItems({QStringLiteral("æ ¡å›­è“"), QStringLiteral("ç»¿æ„"), QStringLiteral("å¤œé—´")});
    navLayout->insertWidget(navLayout->count() - 1, new QLabel(QStringLiteral("å¤–è§‚ä¸»é¢˜"), ui->navPanel));
    navLayout->insertWidget(navLayout->count() - 1, skinCombo);
    connect(skinCombo, &QComboBox::currentTextChanged, [this](const QString& skin) {
        if (skin.contains(QStringLiteral("å¤œ"))) {
            this->setStyleSheet("background:#1f1f2e;color:#f0f0f0;");
        } else if (skin.contains(QStringLiteral("ç»¿"))) {
            this->setStyleSheet("background:#f0fff4;");
        } else {
            this->setStyleSheet("");
        }
    });

    auto* moodCombo = new QComboBox(ui->navPanel);
    moodCombo->addItems({QStringLiteral("ğŸ˜Š å¼€å¿ƒ"), QStringLiteral("ğŸ˜ å¹³é™"), QStringLiteral("ğŸ˜¢ ä½è½")});
    navLayout->insertWidget(navLayout->count() - 1, new QLabel(QStringLiteral("ä»Šæ—¥å¿ƒæƒ…"), ui->navPanel));
    navLayout->insertWidget(navLayout->count() - 1, moodCombo);
    connect(moodCombo, &QComboBox::currentTextChanged, [this](const QString& mood) {
        rove::logging::LogEntry::MoodTag tag = rove::logging::LogEntry::MoodTag::Neutral;
        if (mood.contains(QStringLiteral("å¼€å¿ƒ"))) {
            tag = rove::logging::LogEntry::MoodTag::Happy;
        } else if (mood.contains(QStringLiteral("ä½è½"))) {
            tag = rove::logging::LogEntry::MoodTag::Sad;
        }
        m_logManager.recordManualLog(mood.toStdString(), tag);
        showRealtimeNotification(QStringLiteral("å·²è®°å½•å¿ƒæƒ…ï¼š%1").arg(mood));
    });

    auto* couponBtn = new QPushButton(QStringLiteral("ä½¿ç”¨å®½æ•åˆ¸"), ui->navPanel);
    navLayout->insertWidget(navLayout->count() - 1, couponBtn);
    connect(couponBtn, &QPushButton::clicked, [this] {
        const auto username = m_userManager.activeUser().username();
        bool ok = m_inventoryManager.consumeEffectToken(username, rove::data::ShopItem::PropEffectType::ForgivenessCoupon);
        showRealtimeNotification(ok ? QStringLiteral("å®½æ•åˆ¸å·²ä½¿ç”¨ï¼Œä¸‹ä¸€æ¬¡å¤±è´¥å…æƒ©ç½š")
                                    : QStringLiteral("æ²¡æœ‰å¯ç”¨çš„å®½æ•åˆ¸"));
    });

    // èƒŒåŒ…ç®€æ˜“è¡¨æ ¼
    auto* inventoryTable = new QTableWidget(ui->navPanel);
    inventoryTable->setColumnCount(3);
    inventoryTable->setHorizontalHeaderLabels({QStringLiteral("åç§°"), QStringLiteral("æ•°é‡"), QStringLiteral("å±æ€§")});
    navLayout->insertWidget(navLayout->count() - 1, new QLabel(QStringLiteral("éšèº«é“å…·"), ui->navPanel));
    navLayout->insertWidget(navLayout->count() - 1, inventoryTable);
    auto refreshInventory = [this, inventoryTable]() {
        const auto username = m_userManager.activeUser().username();
        const auto items = m_inventoryManager.listByOwner(username);
        inventoryTable->setRowCount(static_cast<int>(items.size()));
        int row = 0;
        for (const auto& it : items) {
            inventoryTable->setItem(row, 0, new QTableWidgetItem(QString::fromStdString(it.name())));
            inventoryTable->setItem(row, 1, new QTableWidgetItem(QString::number(it.quantity())));
            inventoryTable->setItem(row, 2, new QTableWidgetItem(QString::fromStdString(it.specialAttributes())));
            ++row;
        }
    };
    refreshInventory();
    connect(m_shopInterface, &ShopInterface::purchaseRequested, this, [refreshInventory]() { refreshInventory(); });
}

void MainWindow::connectSignals() {
    connect(m_taskView, &TaskView::taskCompletionRequested, this, [this](int taskId) {
        m_taskManager.markTaskCompleted(taskId);
        m_logManager.recordAutoLog(rove::logging::LogEntry::LogType::Task,
                                   QStringLiteral("ä»»åŠ¡å®Œæˆ").toStdString(),
                                   {});
        m_tutorialManager->markStepDone(QStringLiteral("createTask"));
        refreshDashboard();
    });

    connect(m_shopInterface, &ShopInterface::purchaseRequested, this, [this](int itemId) {
        auto result = m_shopManager.purchaseItem(itemId, 1);
        showRealtimeNotification(QString::fromStdString(result.message));
        m_tutorialManager->markStepDone(QStringLiteral("firstPurchase"));
        refreshDashboard();
    });

    connect(m_customizationPanel,
            &CustomizationPanel::customAchievementCreated,
            this,
            [this](const QString&, const QString&) { m_tutorialManager->markStepDone(QStringLiteral("firstAchievement")); });

    connect(&m_achievementManager,
            &rove::data::AchievementManager::achievementUnlocked,
            this,
            [this](int) { showRealtimeNotification(QStringLiteral("æ–°çš„æˆå°±å·²è§£é”ï¼")); });

    connect(m_userManager.signalProxy(), &rove::data::UserManager::SignalProxy::coinsChanged, this, [this](int) {
        refreshDashboard();
    });

    connect(m_tutorialManager, &TutorialManager::tutorialHintChanged, this, &MainWindow::showRealtimeNotification);
    connect(m_tutorialManager, &TutorialManager::tutorialFinished, this, &MainWindow::handleTutorialFinished);

    // ä»»åŠ¡æé†’å®šæ—¶å™¨
    m_reminderTimer = new QTimer(this);
    m_reminderTimer->setInterval(600000);
    connect(m_reminderTimer, &QTimer::timeout, this, [this] {
        showRealtimeNotification(QStringLiteral("è®°å¾—æŸ¥çœ‹ä»Šæ—¥ä»»åŠ¡ï¼Œä¿æŒæˆé•¿èŠ‚å¥ï¼"));
    });
    m_reminderTimer->start();

    // è§¦å‘ä¸€æ¬¡å¥‡é‡
    auto result = m_serendipityEngine.triggerDailyLogin();
    if (result.triggered) {
        showRealtimeNotification(QString::fromStdString(result.description));
    }
}

void MainWindow::setupTrayIcon() {
    m_trayIcon = new QSystemTrayIcon(this);
    m_trayIcon->setToolTip(QStringLiteral("å…°å¤§æˆé•¿æ¨¡æ‹Ÿ"));
    m_trayIcon->show();
}

void MainWindow::onSectionChanged(int index) {
    ui->stackedWidget->setCurrentIndex(index);
}

void MainWindow::showRealtimeNotification(const QString& message) {
    ui->notificationLabel->setText(message);
    if (m_trayIcon) {
        m_trayIcon->showMessage(QStringLiteral("æ ¡å›­æé†’"), message);
    }
}

void MainWindow::refreshDashboard() {
    if (!m_userManager.hasActiveUser()) {
        return;
    }
    const auto& user = m_userManager.activeUser();
    m_dashboard->renderUser(user);
    m_achievementGallery->reload();
    m_taskView->reloadTasks();
    m_shopInterface->reload();
    m_logBrowser->reload();

    const auto snapshots = m_logManager.querySnapshots(std::nullopt, std::nullopt);
    m_growthDashboard->render(user, snapshots);
}

void MainWindow::handleTutorialFinished() {
    showRealtimeNotification(QStringLiteral("æ•™ç¨‹å®Œæˆï¼Œå¥–åŠ±é‡‘å¸å·²å‘æ”¾"));
}
